File: SUMO-Toolbox-7.0/src/java/ibbt/sumo/examples/HumanSimulator.java
package ibbt.sumo.examples;
/**----------------------------------------------------------------------------------------
** This file is part of the Surrogate Modeling Toolbox ("SUMO Toolbox")
**
** This program is free software; you can redistribute it and/or modify it under
** the terms of the GNU Affero General Public License version 3 as published by the
** Free Software Foundation.
** 
** This program is distributed in the hope that it will be useful, but WITHOUT ANY
** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
** PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
** 
** You should have received a copy of the GNU Affero General Public License along
** with this program; if not, see http://www.gnu.org/licenses or write to the Free
** Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
** 02110-1301 USA, or download the license from the following URL:
** 
** http://www.sumo.intec.ugent.be
** 
** In accordance with Section 7(b) of the GNU Affero General Public License, these
** Appropriate Legal Notices must retain the display of the "SUMO Toolbox" text and
** homepage.  In addition, when mentioning the program in written work, reference
** must be made to the corresponding publication.
** 
** You can be released from these requirements by purchasing a commercial license.
** Buying such a license is in most cases mandatory as soon as you develop
** commercial activities involving the SUMO Toolbox software. Commercial activities
** include: consultancy services or using the SUMO Toolbox in commercial projects 
** (standalone, on a server, through a webservice or other remote access technology).
** 
** For more information, please contact SUMO lab at
** 
**             sumo@intec.ugent.be - www.sumo.intec.ugent.be
**
** Revision: $Id: HumanSimulator.java 6376 2009-12-11 10:13:11Z dgorissen $
**-----------------------------------------------------------------------------------------
*/

import ibbt.sumo.sampleevaluators.EvaluationUnitBatch;
import ibbt.sumo.sampleevaluators.SampleEvaluatorException;
import ibbt.sumo.sampleevaluators.SamplePoint;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Simulator where a human fills in the responses for the requested points
 */
public class HumanSimulator  extends AbstractSimulator{
	
	private static Logger logger = Logger.getLogger("ibbt.sumo.examples.HumanSimulator");
	
	public void simulateBatch(EvaluationUnitBatch batch) throws SampleEvaluatorException {

		System.out.println("===============================================");
		System.out.println("=== Starting HumanSimulator");
		System.out.println("===============================================");
		System.out.println();
		System.out.println("** The SUMO Toolbox is requesting " + batch.getSamples().length + " samples: ");
		System.out.println();
		System.out.println("** Please enter the simulator outputs. If there are multiple outputs, " +
				"numbers should be space separated.  Complex numbers should be specified as two (real/imag)." +
				"\nType RESTART to start again.");
		System.out.println();
		
		int i = 1;
		for (SamplePoint point : batch.getSamples()) {
			System.out.println(i + ") " + point.inputsToString() + " (id=" + point.getId() +")");
		}
		System.out.println();
		
		super.simulateBatch(batch);
		
		System.out.println("===============================================");
	}
	
	public void simulate(SamplePoint point) {
		int maxTries = 10;
		int k = 0;
		while(k < maxTries){
			++k;
			String line = "";
		
			System.out.println("* Enter " + point.getOutputDimension() + " output value(s) for point " + point.inputsToString() + " (id=" + point.getId() +") : ");
			
		    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		    try {
		    	//read the line
		        line = br.readLine();
		        line = line.trim();

		        //start again
		        if(line.equalsIgnoreCase("RESTART")){
		        	continue;
		        }
		        
		        //split into the different outputs
		        Scanner scanner = new Scanner(line);
		        point.outputsFromStream(scanner);
		        
		        //all ok
		        break;
		        
		    } catch (Exception e) { 
		    	logger.severe("Problem while reading output values: " + e.getMessage() + ", please try again.");
		    	e.printStackTrace();
		    	logger.log(Level.FINE,e.getMessage(),e);
		    	continue;
		    }

		}
		
		if(k > maxTries){
		   	logger.severe("There seem to have been problems reading the outputs, considering the point as failed..");
		}
	}
}

File: SUMO-Toolbox-7.0/src/matlab/tools/humanSimulator.m
function out = humanSimulator(varargin)

% humanSimulator (SUMO)
%
%     This file is part of the Surrogate Modeling Toolbox ("SUMO Toolbox")
%     and you can redistribute it and/or modify it under the terms of the
%     GNU Affero General Public License version 3 as published by the
%     Free Software Foundation.  With the additional provision that a commercial
%     license must be purchased if the SUMO Toolbox is used, modified, or extended
%     in a commercial setting. For details see the included LICENSE.txt file.
%     When referring to the SUMO-Toolbox please make reference to the corresponding
%     publication.
%
% Contact : sumo@intec.ugent.be - www.sumo.intec.ugent.be
% Revision: $Rev: 6376 $
%
% Signature:
%	out = humanSimulator(varargin)
%
% Description:
%	An interactive simulator that asks for responses interactively

in = [varargin{:}];

disp('===============================================');
disp('=== Starting HumanSimulator');
disp('===============================================');
disp(' ');
disp(sprintf('** The SUMO Toolbox is requesting %d points:', + size(in,1)));

in

disp(sprintf('** Please enter the simulator outputs. If there are multiple outputs, numbers should be space separated.\nComplex numbers should be specified as two (real/imag).\nType RESTART to start again.'));
disp(' ');

disp('===============================================');

out = [];
i = 1;
while(i <= size(in,1))
	line = input(sprintf('* Enter the output(s) for point %s (type RESTART to start again): ',arr2str(in(i,:))),'s');
	
	if(strcmp(line,'RESTART'))
		out = [];
		i = 1;
        disp('  RESTARTING....');
	end
	
	try
		tmp = str2num(line);
		out = [out ; tmp];
		i = i + 1;
		disp(sprintf('	-> read output(s) %s',arr2str(tmp)));
		
		if(i > size(in,1))
			disp(sprintf(' ==> %d points read : '));
			points = [in out]
			reply = input('Is the above correct? If so type yes, else type RESTART : ','s');

			if(strcmp(reply,'yes'))
				% do nothing
			else
				out = [];
				i = 1;
                disp('  RESTARTING....');
			end
		end
	catch ME
		disp(sprintf('Failed to read outputs: %s, please restart..',ME.message))
		out = [];
		i = 1;
        disp('  RESTARTING....');
	end
end

disp('===============================================');
